examples/vision/image_captioning.py

```diff
@@ -0,0 +1,486 @@
+        grads = tape.gradient(loss, train_vars)
+
+        # 8. Update the trainable weights
+        self.optimizer.apply_gradients(zip(grads, train_vars))
+
+        # 9. Update the loss tracker and return the loss value
+        self.train_loss_tracker.update_state(loss)
+        return {"loss": self.train_loss_tracker.result()}
+
+    def test_step(self, batch_data):
+        loss = self._compute_loss(batch_data, training=False)
+        self.val_loss_tracker.update_state(loss)
+        return {"loss": self.val_loss_tracker.result()}
+
+    @property
+    def metrics(self):
+        # We need to list our metrics here so the `reset_states()` can be
+        # called automatically.
+        return [self.train_loss_tracker, self.val_loss_tracker]
+
+
+cnn_model = get_cnn_model()
+encoder = TransformerEncoderBlock(
+    embed_dim=EMBED_DIM, dense_dim=FF_DIM, num_heads=NUM_HEADS
+)
+decoder = TransformerDecoderBlock(
+    embed_dim=EMBED_DIM, ff_dim=FF_DIM, num_heads=NUM_HEADS
+)
+caption_model = ImageCaptioningModel(
+    cnn_model=cnn_model, encoder=encoder, decoder=decoder
+)
+
+"""
+## Model training
+"""
+
+# Define the loss function
+ce = keras.losses.SparseCategoricalCrossentropy(from_logits=False)
+
+# EarlyStopping criteria
+es = keras.callbacks.EarlyStopping(patience=3, restore_best_weights=True)
+
+# Compile the model
+caption_model.compile(optimizer=keras.optimizers.Adam(), loss=ce)
+
+# Fit the model
+caption_model.fit(
+    train_dataset, epochs=20, validation_data=valid_dataset, callbacks=[es]
+)
+
+"""
+## Check sample predictions
+"""
+
+vocab = vectorization.get_vocabulary()
+index_lookup = dict(zip(range(len(vocab)), vocab))
+max_decoded_sentence_length = MAX_SEQ_LENGTH - 1
+valid_images = list(valid_data.keys())
+
+
+def generate_caption():
+    # Select a random image from the validation dataset
+    sample_img = np.random.choice(valid_images)
+
+    # Read the image from the disk
+    sample_img = read_image(sample_img)
+    img = sample_img.numpy().astype(np.uint8)
+    plt.imshow(img)
+    plt.show()
+
+    # Pass the image to the CNN
+    img = tf.expand_dims(sample_img, 0)
+    img = cnn_model(img)
+
+    # Pass the image features to the Transformer encoder
+    encoded_img = encoder(img, training=False)
+
+    # Generate the caption using the Transformer decoder
+    decoded_caption = "<start> "
+    for i in range(max_decoded_sentence_length):
+        tokenized_target_sentence = vectorization([decoded_caption])[:, :-1]
+        predictions = decoder(tokenized_target_sentence, encoded_img, training=False)
+        sampled_token_index = np.argmax(predictions[0, i, :])
+        sampled_token = index_lookup[sampled_token_index]
+        if sampled_token == " <end>":
+            break
+        decoded_caption += " " + sampled_token
+
+    print("PREDICTED CAPTION:", end=" ")
+    print(decoded_caption.replace("<start> ", "").replace(" <end>", "").strip())
+
+
+# Check predictions for a few samples
```

Comment:
You may want to include a text block that features hardcoded example outputs, along with a comment saying that this is what we get after X epochs. It's often the case that we have to reduce the number of epochs in the example to keep the code easily runnable.

Reply:
Okay. I will add it